addons/udes_stock/models/stock_move_line.py

```diff
@@ -332,6 +332,84 @@ def prepare(
                 quantity_fulfilled = 0
                 if is_serial:
                     qty_done = 1
+
+                # Swapping the tracked products
+                lot_names_on_picking = self.lot_id.mapped("name")
+                lot_names_swapped_in = list(set(lot_names) - set(lot_names_on_picking))
+                # If there swapping tracked products are allowed and some lot_names have been swapped in
+                if mls.u_picking_type_id.u_allow_swapping_tracked_products and lot_names_swapped_in:
+                    # Find movelines in picking for location scanned in
+                    picking_mls, new_ml = mls._find_move_lines(quantity, product, None, None, location)
+                    picking_mls = picking_mls.filtered(lambda m: m.qty_done < m.product_uom_qty)
+                    mls_need_changing = picking_mls.filtered(lambda mls: mls.lot_id.name not in lot_names) 
+                    quantity_swapping = sum(mls_need_changing.mapped("product_uom_qty"))
+
+                    # Workout the lot_ids that were swapped in and store the ones that were swapped out
+                    lot_ids_swapped_in = self.env["stock.production.lot"].search([("name", "in", lot_names_swapped_in)])
+                    lot_ids_swapped_out = picking_mls.lot_id
+                    
+                    # If any of the lot_ids_swapped_in belong to another move_line
+                    swapped_lot_on_assigned_mls = self.env["stock.move.line"].search([("lot_id", "in", lot_ids_swapped_in.mapped("id"))])
+                
+                    # Generator expressions to easily iterate over
+                    mls_need_changing = (ml for ml in mls_need_changing)
+                    lot_ids_swapped_in = (lot for lot in lot_ids_swapped_in)
+                    
+                    if is_serial:
+                        # Want to update the movelines on each serial tracked product
+                        # until all the quantity scanned in is zero
+                        while quantity_swapping > 0:
+                            ml, lot_id = next(mls_need_changing), next(lot_ids_swapped_in)
+                            prod_dict = {
```

Comment:
Q: Indentation looks a bit off, has the file been black formatted?

Reply:
it had not - but they all have been now